k8scrdclient/k8scrdclient.go

```diff
@@ -120,24 +122,26 @@ func (c *CRDClient) ensureCreated(ctx context.Context, crd *apiextensionsv1beta1
 	return nil
 }
 
-// ensureStatusSubresourceCreated ensures if the CRD has a status subresource
-// it is created. This is needed if a previous version of the CRD without the
-// status subresource is present.
-func (c *CRDClient) ensureStatusSubresourceCreated(ctx context.Context, crd *apiextensionsv1beta1.CustomResourceDefinition, b backoff.Interface) error {
-	if crd.Spec.Subresources == nil || crd.Spec.Subresources.Status == nil {
-		// Nothing to do.
-		return nil
-	}
-
+// ensureUpdated ensures if the CRD changed it is updated accordingly. This is
+// needed if e.g. a previous version of the CRD without the status subresource
+// is present where it should actually be set. Another example would be the CRD
+// apiversion changing, which tends to happen every now and then over the
+// runtime object lifecycle and community adoption.
+func (c *CRDClient) ensureUpdated(ctx context.Context, desired *apiextensionsv1beta1.CustomResourceDefinition, b backoff.Interface) error {
 	o := func() error {
-		manifest, err := c.k8sExtClient.ApiextensionsV1beta1().CustomResourceDefinitions().Get(crd.Name, metav1.GetOptions{})
+		current, err := c.k8sExtClient.ApiextensionsV1beta1().CustomResourceDefinitions().Get(desired.Name, metav1.GetOptions{})
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		latest, err := crdVersionLatest(desired, current)
 		if err != nil {
 			return microerror.Mask(err)
 		}
+		if latest {
```

Comment:
I think we also need to check if there are any changes that should be applied. Need to check if there is some functionality to compare the schemas. In worst case we be good with some stupid string comparison.

Reply:
Added some stupid string comparison which should be good enough. Will test.